src/arch/src/aarch64/gic/regs/mod.rs

```diff
@@ -0,0 +1,206 @@
+mod dist_regs;
+mod icc_regs;
+mod redist_regs;
+
+use crate::aarch64::gic::{Error, Result};
+use kvm_bindings::kvm_device_attr;
+use kvm_ioctls::DeviceFd;
+use std::iter::StepBy;
+use std::ops::Range;
+use versionize::{VersionMap, Versionize, VersionizeResult};
+use versionize_derive::Versionize;
+
+pub trait MmappedReg {
+    fn range(&self) -> Range<u64>;
+
+    fn iter<T>(&self) -> StepBy<Range<u64>>
+    where
+        Self: Sized,
+    {
+        self.range().step_by(std::mem::size_of::<T>())
+    }
+}
+
+pub trait VgicRegEngine {
+    type Reg: MmappedReg;
+    type RegChunk: Default + Clone;
+
+    fn group() -> u32;
+
+    fn mpidr_mask() -> u64;
+
+    fn kvm_device_attr(offset: u64, val: &mut Self::RegChunk, mpidr: u64) -> kvm_device_attr {
+        kvm_device_attr {
+            group: Self::group(),
+            attr: (mpidr & Self::mpidr_mask()) | offset,
+            addr: val as *mut Self::RegChunk as u64,
+            flags: 0,
+        }
+    }
+
+    #[inline]
+    fn get_reg_data(fd: &DeviceFd, reg: &Self::Reg, mpidr: u64) -> Result<Vec<Self::RegChunk>>
+    where
+        Self: Sized,
+    {
+        let mut data = Vec::with_capacity(reg.iter::<Self::RegChunk>().count());
+        for offset in reg.iter::<Self::RegChunk>() {
+            let mut val = Self::RegChunk::default();
+            fd.get_device_attr(&mut Self::kvm_device_attr(offset, &mut val, mpidr))
+                .map_err(|e| Error::DeviceAttribute(e, false, Self::group()))?;
+            data.push(val);
+        }
+
+        Ok(data)
+    }
+
+    fn get_regs_data(
+        fd: &DeviceFd,
+        regs: Box<dyn Iterator<Item = &Self::Reg>>,
+        mpidr: u64,
+    ) -> Result<Vec<Vec<Self::RegChunk>>>
+    where
+        Self: Sized,
+    {
+        let mut data = Vec::new();
```

Comment:
nit:
```suggestion
        let mut data = Vec::with_capacity(regs.len());
```

Reply:
I think above comment was meant for another thread 😛

Reply:
I double checked. It also happens here.

Reply:
oooh, getting `len()` using an iterator implies actually running the iterator to its end, which obviously consumes it.

`Vec::with_capacity(regs_iter.clone().len());` should work if the underlying iter supports clone.

This is only a nit anyway, the create snapshot performance is not critical.